x/expiration/keeper/grpc_query.go

```diff
@@ -0,0 +1,141 @@
+package keeper
+
+import (
+	"context"
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/query"
+
+	"github.com/provenance-io/provenance/x/expiration/types"
+)
+
+var _ types.QueryServer = Keeper{}
+
+// Params queries params
+func (k Keeper) Params(
+	goCtx context.Context,
+	_ *types.QueryParamsRequest,
+) (*types.QueryParamsResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	var params types.Params
+	k.paramSpace.GetParamSet(ctx, &params)
+
+	return &types.QueryParamsResponse{Params: params}, nil
+}
+
+// Expiration queries for an expiration based on request parameters
+func (k Keeper) Expiration(
+	goCtx context.Context,
+	req *types.QueryExpirationRequest,
+) (*types.QueryExpirationResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	expiration, err := k.GetExpiration(ctx, req.ModuleAssetId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.QueryExpirationResponse{Expiration: expiration}, nil
+}
+
+// AllExpirations queries for all expirations
+func (k Keeper) AllExpirations(
+	goCtx context.Context,
+	req *types.QueryAllExpirationsRequest,
+) (*types.QueryAllExpirationsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pageRes, expirations, err := k.filteredPaginate(ctx, req.Pagination, func(expiration types.Expiration) bool {
+		return true // do not filter
+	})
+
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &types.QueryAllExpirationsResponse{Expirations: expirations, Pagination: pageRes}, nil
+}
+
+// AllExpirationsByOwner queries all expirations for a particular owner
+func (k Keeper) AllExpirationsByOwner(
+	goCtx context.Context,
+	req *types.QueryAllExpirationsByOwnerRequest,
+) (*types.QueryAllExpirationsByOwnerResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pageRes, expirations, err := k.filteredPaginate(ctx, req.Pagination, func(expiration types.Expiration) bool {
+		return strings.TrimSpace(req.Owner) == expiration.Owner
```

Comment:
Yes, we do need to and an index. I did not attempt to add it in on this iteration of the module. I felt it would be best to add it later on after a first release was out. I created an issue for it #1123